File: backend/app/schemas/rag_chat.py
"""
RAG Chat 请求/响应模型。

用于 Self-RAG 问答功能的 Pydantic 模型定义。
"""

from typing import List, Optional, Literal
from pydantic import BaseModel, Field


class ChatMessage(BaseModel):
    """单条对话消息"""
    role: Literal["user", "assistant"] = Field(..., description="消息角色")
    content: str = Field(..., min_length=1, description="消息内容")


class RagChatRequest(BaseModel):
    """RAG Chat 请求"""
    messages: List[ChatMessage] = Field(..., min_length=1, description="对话历史")
    top_k: int = Field(default=10, ge=1, le=30, description="检索文档数量")
    min_score: float = Field(default=0.3, ge=0.0, le=1.0, description="最小相似度阈值")


class RetrievedSource(BaseModel):
    """检索到的来源"""
    id: str
    index: int = Field(..., ge=1, description="来源索引，从1开始，用于引用标记")
    content: str = Field(..., description="内容片段")
    score: float = Field(..., ge=0.0, le=1.0, description="相似度分数")
    source_type: Literal["article", "repository"] = Field(..., description="来源类型")
    title: str = Field(..., description="标题")
    url: Optional[str] = Field(None, description="链接")
    # Repository 专用字段（用于引用卡片显示）
    owner_login: Optional[str] = Field(None, description="仓库所有者用户名")
    owner_avatar_url: Optional[str] = Field(None, description="所有者头像URL")
    stargazers_count: Optional[int] = Field(None, description="Star数量")
    language: Optional[str] = Field(None, description="主要编程语言")
    description: Optional[str] = Field(None, description="仓库描述")


class RagChatResponse(BaseModel):
    """RAG Chat 非流式响应（备用）"""
    answer: str
    sources: List[RetrievedSource]
    needs_retrieval: bool
    supported: bool
    utility: float

File: frontend/lib/api/rag-chat.ts
/**
 * RAG Chat API 客户端
 *
 * 处理 SSE 流式响应
 */

import { fetchWithAuth, isTokenExpiringSoon, proactiveRefresh } from "./fetch-client"

const API_BASE = "/api/backend/rag-chat"

export interface ChatMessage {
  role: "user" | "assistant"
  content: string
}

export interface RetrievedSource {
  id: string
  index: number // 来源索引，从1开始，用于引用标记
  content: string
  score: number
  source_type: "article" | "repository"
  title: string
  url?: string
  // Repository 专用字段（用于引用卡片显示）
  owner_login?: string
  owner_avatar_url?: string
  stargazers_count?: number
  language?: string
  description?: string
}

export type StreamEventType =
  | "decision"
  | "retrieval"
  | "content"
  | "assessment"
  | "done"
  | "error"

export interface StreamEvent {
  event: StreamEventType
  data: Record<string, unknown>
}

export interface RagChatRequest {
  messages: ChatMessage[]
  top_k?: number
  min_score?: number
}

export const ragChatApi = {
  /**
   * 流式问答
   */
  async streamChat(
    messages: ChatMessage[],
    onEvent: (event: StreamEvent) => void,
    signal?: AbortSignal
  ): Promise<void> {
    // Proactive refresh before SSE long-running request
    if (isTokenExpiringSoon()) {
      const refreshed = await proactiveRefresh()
      if (!refreshed) {
        throw new Error("Session expired")
      }
    }

    const response = await fetchWithAuth(`${API_BASE}/stream`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ messages }),
      signal,
    })

    if (!response.ok) {
      const error = await response.json().catch(() => ({}))
      throw new Error(error.detail || `Request failed: ${response.status}`)
    }

    const reader = response.body?.getReader()
    if (!reader) throw new Error("No response body")

    const decoder = new TextDecoder()
    let buffer = ""
    let currentEvent = ""

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split("\n")
      buffer = lines.pop() || ""

      for (const line of lines) {
        if (line.startsWith("event: ")) {
          currentEvent = line.slice(7).trim()
        } else if (line.startsWith("data: ")) {
          try {
            const data = JSON.parse(line.slice(6))
            onEvent({ event: currentEvent as StreamEventType, data })
          } catch (e) {
            console.error("Failed to parse SSE data:", e)
          }
        }
      }
    }
  },
}
